Feature/4051 sequence diagram multi directional arrow

Ronid1 opened the pull request:
## :bookmark_tabs: Summary

Added Bi-directional arrow support for dotted and solid lines in sequence diagram

Resolves #4051 

## :straight_ruler: Design Decisions

Added new syntax ``<<->>`` for solid bidirectional arrow
Added new syntax ``<<-->>`` for dotted bidirectional arrow
Please approve new syntax.

Added unit tests for both arrows.

### :clipboard: Tasks

Make sure you

- [X] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [X] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/contributing.md#update-documentation) is used for all new features.
- [X] :bookmark: targeted `develop` branch

sidharthv96 commented on packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:
```diff
@@ -516,6 +516,34 @@ Alice->>Bob:Hello Bob, how are you?`;
+    expect(messages[0].type).toBe(diagram.db.LINETYPE.BIDIRECTIONAL_SOLID);
+  });
+  it('should handle bidirectional dotted arrow messages', async () => {
+    const str = 'sequenceDiagram\n' + 'Alice--><Bob:Hello Bob, how are you?';
```
Please use backticked strings like above, so it's more readable.

sidharthv96 reviewed (changes requested):
LGTM! Nice work :)

Please add corresponding rendering tests as well in `cypress/integration/rendering/sequencediagram.spec.js`

Ronid1 commented:
@sidharthv96 I added e2e test and modified the syntax, since the `<` char was parsed as an HTML open tag.
Updated new syntax `->*>` for solid bidirectional arrow
Updated new syntax `-->*>` for dotted bidirectional arrow

Should I add documentation in a separate PR?

sidharthv96 commented on cypress/integration/rendering/sequencediagram.spec.js:
```diff
@@ -68,6 +68,21 @@ context('Sequence diagram', () => {
+      John-->*>Alice: Yes
+      Alice->John: Test
+      John->>Alice: Still works
+      `,
+      { logLevel: 0 }
+    );
+    cy.get('svg');
```
```suggestion
      `
    );
```

sidharthv96 reviewed (approved):
Technically, the changes are perfect. 
The only concern is the proposed syntax. It's unfortunate that we can't use `<`.

Is `-*>` and `--*>` better, so that we don't go into 5 character arrows?

Ronid1 commented:
Agree that `-*>` and `--*>` are better. updated.

sidharthv96 commented:
https://mermaid.js.org/syntax/flowchart.html#multi-directional-arrows

We are using `<-->` for multidimensional arrows. And it'll be supported in markdown usecases, and live editor. 
The only problem is when we write an HTML page directly no?

Ronid1 commented:
@sidharthv96 would love to make progress with this PR, could you please help me tag the correct people for the second review you requested?

sidharthv96 commented:
`<<->>` and `<<-->>` are the clear choice from a uniformity and intutiveness perspective, and feels like a worthy tradeoff against line length. If someone is editing a diagram with a `-->>`, and wants to add a left arrow, the natural thing to try is `<<-->>`, unless they consult the doc and have the syntax in memory.

I know we've gone back and forth on this for a while, but this seems like the best option for our users. Initially I thought we could never use `<`, but as that's not the case, I'm good with this option. 

@mermaid-js/atlantians thoughts?

knsv commented:
@Ronid1 Thanks for adding this feature. One of the challenges for Mermaid is to keep the syntaxes coherent between the different diagrams. The proposed syntax offers better readability and stays pretty consistent with the flowchart syntax for multi-directional arrows.

sidharthv96 commented:
@Ronid1 can you please resolve the conflicts?

Ronid1 commented:
@sidharthv96 should be good to go now

sidharthv96 commented:
Just noticed that the arrow is overlapping the box by a few pixels? Can you fix that?

Proper position
<img width="86" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/7dd4dd2e-2d65-44ed-88a2-e20eb657ff2e">

Incorrect
<img width="132" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/57c64fc7-5604-413c-937c-f202055843b1">

Ronid1 commented:
@sidharthv96 arrow pointer position fixed

Ronid1 commented:
@sidharthv96 bumping

sidharthv96 commented on packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts:
```diff
@@ -425,7 +425,14 @@ const drawMessage = async function (diagram, msgModel, lineStartY: number, diagO
   } else {
     line = diagram.append('line');
-    line.attr('x1', startx);
+    let adjustedStartx = startx;
+    if (
+      type === diagObj.db.LINETYPE.BIDIRECTIONAL_DOTTED ||
+      type === diagObj.db.LINETYPE.BIDIRECTIONAL_SOLID
+    ) {
+      startx < stopx ? (adjustedStartx += 3) : (adjustedStartx -= 3);
+    }
+    line.attr('x1', adjustedStartx);
```
The startx and stopx values are calculated in the `buildMessageModel` function. 
Adding this adjustment here breaks the Separation of concern, as it makes the `drawMessage` function responsible for the adjustment, where it should only have drawn as per the dimensions calculated in `buildMessageModel`.

sidharthv96 replied:
```diff
diff --git a/packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts b/packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts
index 6bb3c654e..525358d90 100644
--- a/packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts
+++ b/packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts
@@ -425,14 +425,7 @@ const drawMessage = async function (diagram, msgModel, lineStartY: number, diagO
     }
   } else {
     line = diagram.append('line');
-    let adjustedStartx = startx;
-    if (
-      type === diagObj.db.LINETYPE.BIDIRECTIONAL_DOTTED ||
-      type === diagObj.db.LINETYPE.BIDIRECTIONAL_SOLID
-    ) {
-      startx < stopx ? (adjustedStartx += 3) : (adjustedStartx -= 3);
-    }
-    line.attr('x1', adjustedStartx);
+    line.attr('x1', startx);
     line.attr('y1', lineStartY);
     line.attr('x2', stopx);
     line.attr('y2', lineStartY);
@@ -1449,7 +1442,7 @@ const buildMessageModel = function (msg, actors, diagObj) {
   const [fromLeft, fromRight] = activationBounds(msg.from, actors);
   const [toLeft, toRight] = activationBounds(msg.to, actors);
   const isArrowToRight = fromLeft <= toLeft;
-  const startx = isArrowToRight ? fromRight : fromLeft;
+  let startx = isArrowToRight ? fromRight : fromLeft;
   let stopx = isArrowToRight ? toLeft : toRight;
 
   // As the line width is considered, the left and right values will be off by 2.
@@ -1488,6 +1481,14 @@ const buildMessageModel = function (msg, actors, diagObj) {
     if (![diagObj.db.LINETYPE.SOLID_OPEN, diagObj.db.LINETYPE.DOTTED_OPEN].includes(msg.type)) {
       stopx += adjustValue(3);
     }
+
+    if (
+      [diagObj.db.LINETYPE.BIDIRECTIONAL_SOLID, diagObj.db.LINETYPE.BIDIRECTIONAL_DOTTED].includes(
+        msg.type
+      )
+    ) {
+      startx -= adjustValue(3);
+    }
   }
 
   const allBounds = [fromLeft, fromRight, toLeft, toRight];
```

Ronid1 replied:
Thanks for pointing out this function, I made the updates @sidharthv96

sidharthv96 commented:
Found this issue while testing the suggested fix

<img width="914" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/d47c6e33-2d25-431f-955c-f784015912f0">

The arrow doesn't show up when autoNumbering is on. 

<img width="931" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/c380e3d6-2f86-4a1a-9c2f-b6470644501a">

And the circle is misaligned, when compared to uni directional arrows.

Ronid1 commented:
I think the auto-numbering should be solved as a separate issue. Seems to me that the approach to fixing that has to do with moving the number position and not with the arrowhead. Happy to open an issue and work on it.

Ronid1 commented:
bumping @sidharthv96
